app/_hub/kong-inc/grpc-web/index.md

```diff
@@ -0,0 +1,154 @@
+
+## Usage
+
+This plugin is intended to be used in a Kong route between an HTTP endpoint and a gRPC service.
+
+Sample configuration via declarative (YAML):
+
+```yaml
+_format_version: "1.1"
+services:
+- protocol: grpc
+  host: localhost
+  port: 9000
+  routes:
+  - protocols:
+    - http
+    paths:
+    - /
+    plugins:
+    - name: grpc-web
+```
+
+Same thing via the Admin API:
+
+```bash
+$ # add the gRPC service
+$ curl -XPOST localhost:8001/services \
+  --data name=grpc \
+  --data protocol=grpc \
+  --data host=localhost \
+  --data port=9000
+
+$ # add an http route
+$ curl -XPOST localhost:8001/services/grpc/routes \
+  --data protocols=http \
+  --data name=web-service \
+  --data paths=/
+
+$ # add the plugin to the route
+$ curl -XPOST localhost:8001/routes/web-service/plugins \
+  --data name=grpc-web
+```
+
+In these examples we don't set any configuration for the plugin.  This minimal setup works for the default varieties of the [gRPC-Web protocol], which use ProtocolBuffer messages either directly in binary or with base64 encoding.  The related `Content-Type` headers are `application/grpc-web` or `application/grpc-web+proto` for binary and `application/grpc-web-text` or `application/grpc-web-text+proto`.
+
+If we wish to use JSON encoding, we have to provide the gRPC specification in a .proto file, which needs to be installed in the Kong node running the plugin.  For example:
+
+```protobuf
+syntax = "proto2";
+
+package hello;
+
+service HelloService {
+  rpc SayHello(HelloRequest) returns (HelloResponse);
+  rpc LotsOfReplies(HelloRequest) returns (stream HelloResponse);
+  rpc LotsOfGreetings(stream HelloRequest) returns (HelloResponse);
+  rpc BidiHello(stream HelloRequest) returns (stream HelloResponse);
+}
+
+message HelloRequest {
+  optional string greeting = 1;
+}
+
+message HelloResponse {
+  required string reply = 1;
+}
+```
+
+The declarative configuration becomes:
+
+```yaml
+_format_version: "1.1"
+services:
+- protocol: grpc
+  host: localhost
+  port: 9000
+  routes:
+  - protocols:
+    - http
+    paths:
+    - /
+    plugins:
+    - name: grpc-web
+      proto: path/to/hello.proto
+```
+
+or via the Admin API:
+
+```bash
+$ # add the plugin to the route
+$ curl -XPOST localhost:8001/routes/web-service/plugins \
+  --data name=grpc-web \
+  --data proto=path/to/hello.proto
+```
+
+With this setup, we can support gRPC-Web/JSON clients using `Content-Type` headers like `application/grpc-web+json` or `application/grpc-web-text+json`.
+
+Additionaly, non-gRPC-Web clients can do simple POST requests with `Content-Type: application/json`.  In this case, the data payloads are simple JSON objects, without gPRC frames.  This allows clients to perform calls without using the gRPC-Web library, but streaming responses are delivered as a continous stream of JSON objects.  It's up to the client to split into separate objects if it has to support multiple response messages.
```

Comment:
I don't understand what this paragraph means. When a user does a POST with `Content-Type: application/json`, what happens?

> the data payloads are simple JSON objects, without gPRC frames

What does "data payloads" mean here? Is it the consumer -> kong request, the kong -> upstream request, the upstream -> kong response, the kong -> user response, or all of them? My best understanding of this (with low trust) is "the JSON request is proxied verbatim to the upstream". Why? Is it just that the plugin isn't activated by this calls (so they are proxied transparently)?

> streaming responses are delivered as a continous stream of JSON objects

Does this mean that if I send a (non-html) TCP request the plugin is not activated and the TCP input will be proxied to the upstream without any transformation? Or I am not understanding something?

Reply:
not sure where the confusion comes from.  Let me try to write this paragraph again.

Reply:
done